### PR TITLE
User story new ruleset

### PR DESCRIPTION
**Pull request by sophialhales**
## Describe Your Changes
- User story new ruleset completed
- User story cancel button completed
- User story deleteById endpoint completed
- User story updated ruleset completed sans save button

## Related Issues
#73 #72 #63 #48

## Type of Change
Please delete options that are not relevant.
- [x] New feature (change that adds functionality)

## Checklist (DOD)
- [x] The code is written and passes all Acceptance criteria's
- [x] Any Zombie code is removed
- [x] Code has proper comment documentation
- [x] The code is unit tested with happy case and passes
- [x] Code is refactored for optimization
- [x] The code has been peer-reviewed by one or more team members who did not work on it.
- [x] The code has no merge conflicts

### If Applicable Checklist
- [x] The code is integration tested and passes (if applicable)
- [x] Code is reviewed and accepted by Product Owner (if applicable)

### CODE REVIEW

File: front-end/src/app/app.component.spec.ts [hunk 1]
```diff
@@ -1,9 +1,48 @@
...
+import { DeleteRuleButtonComponent } from './delete-rule-button/delete-rule-button.component';
+import { MoveDownButtonComponent } from './move-down-button/move-down-button.component';
+import { MoveUpButtonComponent } from './move-up-button/move-up-button.component';
+import { CreateRulesetComponent } from './create-ruleset/create-ruleset.component';
+import { DashboardPathComponent } from './dashboard-path/dashboard-path.component';
+import { RulesetNameComponent } from './ruleset-name/ruleset-name.component';
+import { SaveButtonComponent } from './save-button/save-button.component';
+import { RouterTestingModule } from '@angular/router/testing';
+import { HttpClientModule } from '@angular/common/http';
+
```
Comment (GivenTeacher5) [inline front-end/src/app/app.component.spec.ts:20]: remove extra whitespace

File: front-end/src/app/app.module.ts [hunk 2]
```diff
@@ -1,16 +1,55 @@
...
+import { RulesetNameComponent } from './ruleset-name/ruleset-name.component';
+import { SaveButtonComponent } from './save-button/save-button.component';
+import { DashboardPathComponent } from './dashboard-path/dashboard-path.component';
+import { CreateRulesetComponent } from './create-ruleset/create-ruleset.component';
+import { CancelButtonComponent } from './cancel-button/cancel-button.component';
+import { AppRoutingModule } from './app-routing.module';
+import { RouterModule } from "@angular/router";
+import { HttpClientModule } from '@angular/common/http';
+import { HashLocationStrategy, LocationStrategy } from '@angular/common';
+
```
Comment (GivenTeacher5) [inline front-end/src/app/app.module.ts:24]: remove extra whitespace

File: front-end/src/app/create-ruleset/create-ruleset.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+  });
+
+  it(`clicking "Add New Rule" should increase the amount of rules`, () => {
+    const fixture = TestBed.createComponent(CreateRulesetComponent);
+    const app = fixture.componentInstance;
+    fixture.debugElement.query(By.css('#ruleButton')).nativeElement.click();
+
+    expect(app.ruleset.length).toBe(2);
+  });
+  it('button should be rendered', () => {
```
Comment (GivenTeacher5) [inline front-end/src/app/create-ruleset/create-ruleset.component.spec.ts:62]: elaborate for readability

File: front-end/src/app/dashboard-path/dashboard-path.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+
```
Comment (GivenTeacher5) [inline front-end/src/app/dashboard-path/dashboard-path.component.spec.ts:2]: WHITESPACE

File: front-end/src/app/rules-component/rules-component.component.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import { Component, Input, ViewChildren, OnInit, QueryList, AfterViewInit } from '@angular/core';
+import { ConditionsComponent } from '../conditions/conditions.component';
+@Component({
```
Comment (GivenTeacher5) [inline front-end/src/app/rules-component/rules-component.component.ts:3]: maybe add whitespace but idk sounds extra to me

File: front-end/src/app/rules-component/rules-component.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  it('When "Add Condition" button is clicked, new condition component is rendered on screen', () => {
+    let conditionComponentsArray = fixture.debugElement.nativeElement.querySelectorAll('.conditionsComponent');
+    expect(conditionComponentsArray.length).toBe(1);
+    let addConditionButton = fixture.debugElement.query(By.css('.addConditionButton')).nativeElement;
+    addConditionButton.click();
+    fixture.detectChanges();
+    conditionComponentsArray = fixture.debugElement.nativeElement.querySelectorAll('.conditionsComponent');
+    expect(conditionComponentsArray.length).toBe(2);
+  });
+
```
Comment (GivenTeacher5) [inline front-end/src/app/rules-component/rules-component.component.spec.ts:51]: WHITESPACE

File: front-end/src/app/save-button/save-button.component.spec.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+      providers: [ CreateRulesetComponent ]
+    });
+
+    fixture = TestBed.createComponent(SaveButtonComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+    
+  });
+
+
```
Comment (GivenTeacher5) [inline front-end/src/app/save-button/save-button.component.spec.ts:25]: whitespace

File: front-end/src/app/save-button/save-button.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+import { SaveButtonComponent } from './save-button.component';
+import { By } from '@angular/platform-browser';
+import { HttpClientTestingModule } from '@angular/common/http/testing';
+import { CreateRulesetComponent } from '../create-ruleset/create-ruleset.component';
+
+
```
Comment (GivenTeacher5) [inline front-end/src/app/save-button/save-button.component.spec.ts:7]: whitespace

File: front-end/src/app/save-button/save-button.component.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+import { HttpClient, HttpHeaders } from '@angular/common/http';
+import { ConditionsComponent } from '../conditions/conditions.component';
+
+@Component({
+  selector: 'app-save-button',
+  templateUrl: './save-button.component.html',
+  styleUrls: ['./save-button.component.css']
+})
+
+
```
Comment (GivenTeacher5) [inline front-end/src/app/save-button/save-button.component.ts:12]: whitespace

File: front-end/src/app/sidebar/sidebar.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,49 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+
```
Comment (GivenTeacher5) [inline front-end/src/app/sidebar/sidebar.component.spec.ts:2]: must i say?

File: front-end/src/main.ts [hunk 11]
```diff
@@ -1,3 +1,5 @@
+/// <reference types="@angular/localize" />
+
 import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
```
Comment (GivenTeacher5) [inline front-end/src/main.ts:4]: ....

File: pom.xml [hunk 12]
```diff
@@ -18,6 +18,7 @@
 		<frontend-dir>${project.basedir}/front-end</frontend-dir>
 	</properties>
 	<dependencies>
+
```
Comment (GivenTeacher5) [inline pom.xml:21]: fix it

File: src/main/java/com/accesspoint/rulesengine/exception/BadRequestException.java [hunk 13]
```diff
@@ -0,0 +1,8 @@
+package com.accesspoint.rulesengine.exception;
+
+
```
Comment (GivenTeacher5) [inline src/main/java/com/accesspoint/rulesengine/exception/BadRequestException.java:3]: why

File: src/main/java/com/accesspoint/rulesengine/service/RulesetService.java [hunk 14]
```diff
@@ -1,23 +1,158 @@
...
+            if (rule.getPriority() == 0.0) throw new BadRequestException("Rule priority cannot be 0");
+            if (rule.getEvent_type() == null) throw new BadRequestException("Rule event type cannot be null");
+            if (rule.getConditions() == null || rule.getConditions().isEmpty()) throw new BadRequestException("Conditions cannot be empty");
+            rule.getConditions().stream().forEach(condition -> {
+                if (condition.getFact_type() == null) throw new BadRequestException("Condition fact type cannot be null");
+                if (condition.getValue_type() == null) throw new BadRequestException("Condition value type cannot be null");
+            });
+        });
+
+
```
Comment (GivenTeacher5) [inline src/main/java/com/accesspoint/rulesengine/service/RulesetService.java:88]: whitespace

File: src/main/java/com/accesspoint/rulesengine/service/RulesetService.java [hunk 15]
```diff
@@ -1,23 +1,158 @@
...
+        Ruleset updatedRuleset2 = rulesetRepository.findById(id) //
+                .map(oldRuleset -> {
+                    //TODO: optimize Hibernate select calls
+                    if (!oldRuleset.getName().equals(incomingRuleset.getName())) oldRuleset.setName(incomingRuleset.getName());
+                    if (oldRuleset.getRules() != incomingRuleset.getRules()) {
+                        for (Rule incomingRule : incomingRuleset.getRules()) {
+
```
Comment (GivenTeacher5) [inline src/main/java/com/accesspoint/rulesengine/service/RulesetService.java:95]: whitespace

**Review by GivenTeacher5 — DISMISSED** [review]
looks like 10% whitespace

File: front-end/src/app/app.component.html [hunk 16]
```diff
@@ -1 +1,7 @@
-<h1>Hello, This is Cyclones!</h1>
+<main class="d-flex flex-nowrap" style="height: fit-content; min-height: 100%;">
+    <app-sidebar class="d-flex flex-column flex-shrink-0 p-3 text-bg-dark col-sm-3 col-12 sidebar" style="max-width: 280px; height: 100vh;"></app-sidebar>
+    <div class="container p-4">
+        <router-outlet></router-outlet>
+        <!-- <app-create-ruleset></app-create-ruleset> -->
```
[the remark this answers is not in the document]
Comment (sophialhales) [reply front-end/src/app/app.component.html:5]: done

File: front-end/src/app/app.module.ts (hunk 2 above)
Comment (sophialhales) [reply front-end/src/app/app.module.ts:24]: done

File: front-end/src/app/cancel-button/cancel-button.component.spec.ts [hunk 17]
```diff
@@ -0,0 +1,35 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+import { By } from '@angular/platform-browser'
+import { CancelButtonComponent } from './cancel-button.component';
+
+
```
[the remark this answers is not in the document]
Comment (sophialhales) [reply front-end/src/app/cancel-button/cancel-button.component.spec.ts:5]: done

File: front-end/src/app/create-ruleset/create-ruleset.component.spec.ts (hunk 3 above)
Comment (sophialhales) [reply front-end/src/app/create-ruleset/create-ruleset.component.spec.ts:62]: done

File: front-end/src/app/dashboard-path/dashboard-path.component.spec.ts (hunk 4 above)
Comment (sophialhales) [reply front-end/src/app/dashboard-path/dashboard-path.component.spec.ts:2]: done

File: front-end/src/app/delete-rule-button/delete-rule-button.component.spec.ts [hunk 18]
```diff
@@ -0,0 +1,42 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+
```
[the remark this answers is not in the document]
Comment (sophialhales) [reply front-end/src/app/delete-rule-button/delete-rule-button.component.spec.ts:2]: done

**Review by ColeNelson806 — APPROVED** [review]
She looks mighty fine and pretty. Great job everyone on this code!

**Pull request merged**